### ISSUE TITLE
viomi.vacuum.v3 support for Mi Robot Vacuum-Mop 2 Pro

### ISSUE THREAD

**Issue by Dejv311** [PiotrMachowski/Home-Assistant-custom-components-Xiaomi-Cloud-Map-Extractor#277]
Hi, please add support for this vacuum. Thank you! :)

**Comment by PiotrMachowski** [comment]
It should be supported, does it not work?

**Comment by Dejv311** [comment]
It is not listed among supported versions on the main page (just v6, v7, v8 and v13 are listed). And it does not work for me :(

**Comment by PiotrMachowski** [comment]
This list is definitely not up to date. What does camera image show? It's there anything in logs?

**Comment by Dejv311** [comment]
The camera image shows "Failed to retrieve map from vacuum" on a gray background

HA logs show (3 logs bellow):

```
Logger: custom_components.xiaomi_cloud_map_extractor.camera
Source: custom_components/xiaomi_cloud_map_extractor/camera.py:309
Integration: xiaomi_cloud_map_extractor ([documentation](https://github.com/PiotrMachowski/Home-Assistant-custom-components-Xiaomi-Cloud-Map-Extractor), [issues](https://github.com/PiotrMachowski/Home-Assistant-custom-components-Xiaomi-Cloud-Map-Extractor/issues))
First occurred: 12:37:21 (1 occurrences)
Last logged: 12:37:21

Got exception while fetching the state: {'code': -9999, 'message': 'user ack timeout'}
```

```
Logger: homeassistant.helpers.entity
Source: custom_components/xiaomi_cloud_map_extractor/camera.py:303
Integration: xiaomi_cloud_map_extractor ([documentation](https://github.com/PiotrMachowski/Home-Assistant-custom-components-Xiaomi-Cloud-Map-Extractor), [issues](https://github.com/PiotrMachowski/Home-Assistant-custom-components-Xiaomi-Cloud-Map-Extractor/issues))
First occurred: 14:08:21 (1 occurrences)
Last logged: 14:08:21

Update for camera.xiaomi_cloud_map_extractor fails
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 514, in async_update_ha_state
    await self.async_device_update()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 750, in async_device_update
    await task
  File "/usr/local/lib/python3.9/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/config/custom_components/xiaomi_cloud_map_extractor/camera.py", line 256, in update
    map_name = self._handle_map_name(counter)
  File "/config/custom_components/xiaomi_cloud_map_extractor/camera.py", line 303, in _handle_map_name
    map_name = self._vacuum.map()[0]
  File "/usr/local/lib/python3.9/site-packages/miio/click_common.py", line 179, in _wrap
    return func(self, *args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/miio/integrations/vacuum/roborock/vacuum.py", line 400, in map
    return self.send("get_map_v1")
  File "/usr/local/lib/python3.9/site-packages/miio/device.py", line 106, in send
    return self._protocol.send(
  File "/usr/local/lib/python3.9/site-packages/miio/miioprotocol.py", line 202, in send
    self.__id = payload["id"]
TypeError: byte indices must be integers or slices, not str
```

```
Logger: homeassistant.helpers.entity
Source: helpers/entity.py:745
First occurred: 12:37:24 (154 occurrences)
Last logged: 14:32:35

Update of camera.xiaomi_cloud_map_extractor is taking over 10 seconds
```

**Comment by PiotrMachowski** [comment]
Try adding: 
```yaml
force_api: roidmi
```

**Comment by Dejv311** [comment]
Now camera entity is dead and this is the log:

```
Logger: homeassistant.helpers.entity
Source: custom_components/xiaomi_cloud_map_extractor/roidmi/vacuum.py:14
Integration: xiaomi_cloud_map_extractor ([documentation](https://github.com/PiotrMachowski/Home-Assistant-custom-components-Xiaomi-Cloud-Map-Extractor), [issues](https://github.com/PiotrMachowski/Home-Assistant-custom-components-Xiaomi-Cloud-Map-Extractor/issues))
First occurred: 18:59:04 (14 occurrences)
Last logged: 19:00:05

Update for camera.xiaomi_cloud_map_extractor fails
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 514, in async_update_ha_state
    await self.async_device_update()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 742, in async_device_update
    raise exc
  File "/usr/local/lib/python3.9/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/config/custom_components/xiaomi_cloud_map_extractor/camera.py", line 261, in update
    self._handle_map_data(map_name)
  File "/config/custom_components/xiaomi_cloud_map_extractor/camera.py", line 318, in _handle_map_data
    map_data, map_stored = self._device.get_map(map_name, self._colors, self._drawables, self._texts,
  File "/config/custom_components/xiaomi_cloud_map_extractor/common/vacuum.py", line 28, in get_map
    map_data = self.decode_map(response, colors, drawables, texts, sizes, image_config)
  File "/config/custom_components/xiaomi_cloud_map_extractor/roidmi/vacuum.py", line 14, in decode_map
    unzipped = gzip.decompress(raw_map)
  File "/usr/local/lib/python3.9/gzip.py", line 556, in decompress
    return f.read()
  File "/usr/local/lib/python3.9/gzip.py", line 300, in read
    return self._buffer.read(size)
  File "/usr/local/lib/python3.9/gzip.py", line 487, in read
    if not self._read_gzip_header():
  File "/usr/local/lib/python3.9/gzip.py", line 435, in _read_gzip_header
    raise BadGzipFile('Not a gzipped file (%r)' % magic)
gzip.BadGzipFile: Not a gzipped file (b'x\x9c')
```

**Comment by PiotrMachowski** [comment]
Ok, can you send me raw map file? 
https://github.com/PiotrMachowski/Home-Assistant-custom-components-Xiaomi-Cloud-Map-Extractor#retrieving-map

`piotr.machowski.dev [at] gmail.com`

**Comment by Dejv311** [comment]
Did you receive my email, @PiotrMachowski?

**Comment by PiotrMachowski** [comment]
Yup, I got it. It's waiting in the backlog

**Comment by rolandasr** [comment]
Hello guys. I have the same problem with the map extractor using this model of robot. I downloaded the file and it appeared to be zlib, not gzip. Magic numbers of the file: 78 9C. Thus, I forced it bu using viomi api. However, I get this error:

Update for camera.xiaomi_cloud_map_extractor fails
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 515, in async_update_ha_state
    await self.async_device_update()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 743, in async_device_update
    raise exc
  File "/usr/local/lib/python3.9/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/config/custom_components/xiaomi_cloud_map_extractor/camera.py", line 261, in update
    self._handle_map_data(map_name)
  File "/config/custom_components/xiaomi_cloud_map_extractor/camera.py", line 318, in _handle_map_data
    map_data, map_stored = self._device.get_map(map_name, self._colors, self._drawables, self._texts,
  File "/config/custom_components/xiaomi_cloud_map_extractor/common/vacuum.py", line 28, in get_map
    map_data = self.decode_map(response, colors, drawables, texts, sizes, image_config)
  File "/config/custom_components/xiaomi_cloud_map_extractor/viomi/vacuum.py", line 15, in decode_map
    return MapDataParserViomi.parse(unzipped, colors, drawables, texts, sizes, image_config)
  File "/config/custom_components/xiaomi_cloud_map_extractor/viomi/map_data_parser.py", line 166, in parse
    _LOGGER.debug('rooms: %s', [str(room) for number, room in map_data.rooms.items()])
AttributeError: 'NoneType' object has no attribute 'items'

**Comment by PiotrMachowski** [comment]
@Dejv311 your vacuum model is actually `ijai.vacuum.v3`, not `viomi.vacuum.v3`, so related issue is here: #200 

@rolandasr this is a different issue, it will be fixed in v2.2.0
